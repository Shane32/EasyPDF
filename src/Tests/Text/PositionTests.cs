using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
