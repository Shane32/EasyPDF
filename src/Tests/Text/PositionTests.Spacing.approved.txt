%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 923/Filter/FlateDecode>>stream
x\x9C\x8D\x97Mo\xDB0\x0C\x86\xEF\xFE\x15>\xB6\x87x\xA4l\xCB\xD6n\xEB\xB0(vY\x97\xDB\xD0\xC3\x90\xA6]\x87\xBA]\xBB\x00\xFB\xFB\x93E\x8A\xA4\x82\x06\x15\xFA\xC5\xC8$\xC5<u\xDEW~n.\xB6M\xEF\xDBi\xF4\xED\xF6\xA6\xF9\xB4m\xBE5\xD8B\xFC\xDA`;\xB9\xF8\xEDni\xA0\xBD\x8C?\xBF\x1B\xE8\xB0\xFD\xD7\xFC\xB8\x8E\xD7o\x1AHyW_8x\xB9{=\xCD9\xE8\xC6\x18-\x1C\xB9n\x08\xAD\xC3y\xFD3\xAC\xAF\xD1\xA7\xBF\xBB\xC6a\xDF\x8D\xC8\x8B\xA8\xA9\x98\xEA\xD7\xEBk\xB4\xE1\xD5\x94\xBB\xC9\x1D6\xB9E\xEAK\xCBi;N\xE7!v\xCD\xF7&7\\d\xB4\x87\xB4\xCAM\x96&O\xB4.?\xC77\xB4\xDD\xAD\x94\xDE}\xC6\x16\xE3\x8B[\x05\xB4&N\x1D\xCE\xEDv\xB1\x10`E\xF9\xE3\xEC\xC3\xC3\xFD\xDD\xE3\xB2\x7F<\xBC?\x1F\xE1\xAC\xFD\xBA\xBF=l\xCF\xD1\xC1\xD9\xD3\x9F\xF3\xEB\xED%\xC1\xA6\x11b\x0B\xC1\x03\x86\x8Fs\xFC\xF6R \x84x\x19M:R\x97\x0C\x09\xC7\x94G\xE9\xEB\xEF\xD4%\x05\x02\x89\x96i\x7FJ\xCF\xB3(%3X\x0C3\xA7TI\x9C\xD2(U\xA0\\\xEC\x17\xAAI}\x8C\xAF\xF6/\xC7\x9C\xE2\x9D\x9A\xC7\xE9g{\x1FA\xBE\x91\xA0\xB8\x93@8\xE5t\xA4.\x99S\xDF\x1BN=r\x97\x14\x08'Z\xA6M{\xE5\x94\xBAdNv0/\x9CR%\xDFOP\xCB)\xD6\xD7B\xBA\xF8\xF9w\xFFp\xFF\xB8?\xC64\x0E2\xCD\xE8\xA6q\xE67\x98\x02\xC1\xC4\xCBh\xD2\x91\xBAdL#\x1ALC\xC8\xB0\x83\xC5D\xCB\xB4)*\xA6\xD4%c\xB2\x83\x82)U\x12\xA64J\x15\xA6\x11:\xA8\xFF\xE0]<\x1D\x0EO\xCB1\xA7(py\x9Ci0\x9C&\xCF\xEF0\x05\xC2\x89\x97\xD1\xA4#u\xC9\x9C|0\x9C\xFC\xC4]R \x9Ch9m\xCA\xE9y\x16\xE5d\x07s\xC2)U\x12\xA74\xCA\x9B\x9C\xD0\xCF\x9D\x8B"\x1Ej5\x8A>u\xA7T*\xA8f\x06\xAB\xE2!\x8Bp(t<\xA8\x90\x07\xA3\xE4A\xA5|\x9E\x0C\xAEy\xE4.)\x10\\\xB4\x9C6\xE5\xF4<\x8B\xE2\xB2\x83\xA9\x9A\xA7J\xC2\x15\xAA\xE4\x1C=v\xAEoC_+T\x84\xEBu\xA9B0\xD2\x89V\xD4\x11\xB3\xAAS$\xC0\xF2\x05\xB4%\xC8\xBDD\xD9\xA1\x90v\x10m\x87R\xDC\xC1\xA8;Xy\x87B\xDF\xED\x94`\x14\x1ET\xE2i\xAC\xB7\xE9\x8DS\x07>u\xA9gwJ\xC1\xD0\xA9\xA0\xA2\xB3R\x8F}\xD6z\x8A\x94^\xAFj/%\xC8\xBD\x84\x9E\xB3\x82\x8F.+>EJ\xCF\xA9\xE6\xE7\x12\x99\xCB\xD0+\xA6T\xDD\xA7r\xA6\xD7W)?z\xE8\xFAucW+k\x0C\xF0Ua\xC3A\x85\x16\x07k\x018d\x0F\xA0H\xF1\xEA\x02R\x82\xDCK\xF0X\x9C+\x82\xFC+
+\xE0\x0B\xBC\xBD1\x03\xEA%\xF8\x8A)\x07s\xBCPC\xA0\xB1\xDE\xC6\xE7\xE29l\xBD\x07\xB1V\xEA\xAE\xEE\xEF~\x9D>\x8Fy\x15`\xF4\xD6\x1A\xD0go\xA0H\xF9yu\x07)A\xEE\xA5\xC72k\x10\xEB'\x86{\x8D\x85E\xF0\x05>\x9A\x19\x93\xA0^\xC2\xAF\x98R}\x82\xCA\x99\x9F\xAFs\x8A8\x11\xF61{\xAC\xD5\xBE\xC4\xEF\x84\xF4\xCD`N\x8D\xD6)p\xCEVA\x91\xD2\x9B\xD5,\xA4\x04\xB9\x97\xD0\x9B\xAC_\xE0\x94\x83"\xA57\xA9e\xE4\x12\x99\xCB\x1Em\xED\x94j\x1BT\xCE\xF4\xE6:\xE3\x80\xA1\x9BB\xEAR\x8D\xEE\xA4\xF2\x05\xFB,P<\x0C\x80<@\xF98\x00\xF6y\xC0\xFAF0\xBE\x11
\xDF\x08\xE2\x1B\xA1\xF4\x8D`|#X\xDF\x08\x85o\x98)\xD7X\xE0\x05\xF3h\x00u\xBE\x11\x1B\x0Ek\xE5P\xAB|\xC4\xEFH\xF8\xFE\x03 \xAF\x04\x0B
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001162 00000 n 
0000000015 00000 n 
0000001250 00000 n 
0000001005 00000 n 
0000001315 00000 n 
0000001360 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 